client/components/dashboard/dashboard-service.js

```diff
@@ -547,34 +548,34 @@ DashboardService.prototype.rewriteQuery = function(widget, replaceParams) {
 
   let widgetConfig = widget.model.datasource.config;
 
-  let project_name = this.arrayUtilService_.getFirst([
+  let project_name = /** @type {string} */ (this.arrayUtilService_.getFirst([
       widgetConfig.results.project_id,
       this.current.model.project_id,
-      this.config.default_project], false);
+      this.config.default_project], false));
   if (project_name === null) {
     this.errorService_.addError(ErrorTypes.DANGER, 'Project name not found.');
   }
 
-  let dataset_name = this.arrayUtilService_.getFirst([
+  let dataset_name = /** @type {string} */ (this.arrayUtilService_.getFirst([
       widgetConfig.results.dataset_name,
       this.current.model.dataset_name,
-      this.config.default_dataset], false);
+      this.config.default_dataset], false));
   if (project_name === null) {
     this.errorService_.addError(ErrorTypes.DANGER, 'Dataset name not found.');
   }
 
-  let table_name = this.arrayUtilService_.getFirst([
+  let table_name = /** @type {string} */ (this.arrayUtilService_.getFirst([
       widgetConfig.results.table_name,
       this.current.model.table_name,
-      this.config.default_table], false);
+      this.config.default_table], false));
   if (project_name === null) {
     this.errorService_.addError(ErrorTypes.DANGER, 'Table name not found.');
   }
 
-  let table_partition = this.arrayUtilService_.getFirst([
+  let table_partition = /** @type {!QueryTablePartitioning} */ (this.arrayUtilService_.getFirst([
```

Comment:
This style of casting isn't safe - you're calling .getFirst with required=false, and that can return null. The cast suppresses the warning but you're inviting errors further down the line if it turns out to be an invalid assumption.

I think it would be cleaner to split ArrayUtilService's getFirst into two methods with well-defined return types (throws error or allows null) instead of the opt_required argument, analogous to Pythons array[val] vs array.get[val]. That helps the compiler figure things out without the need for casting, and if the assumption does turn out to be wrong you'll get an exception closer to the source of trouble.

Very optionally, consider looking into generics to support typed containers - not sure if this would be possible in this case though:

https://developers.google.com/closure/compiler/docs/js-for-compiler#generics